### PR TITLE
Add flavor to tspconfig.yaml

### PR DESCRIPTION
**Pull request by timotheeguerin**
## Script used 
```js
// @ts-check

import { globby } from "globby";
import { readFile, writeFile } from "fs/promises";
import { stringify, parseDocument } from "yaml";

const files = await globby("**/tspconfig.yaml");
const emittersToPatch = new Set([
  "@azure-tools/typespec-csharp",
  "@azure-tools/typespec-python",
  "@azure-tools/typespec-ts",
  "@azure-tools/typespec-java",
]);
console.log("Files", files);

for (const file of files) {
  console.log("File", file);
  const content = await readFile(file);
  /** @type {any} */
  const yaml = parseDocument(content.toString());

  let patch = false;

  for (const item of yaml.contents?.get("options").items) {
    if (emittersToPatch.has(item.key.value)) {
      patch = true;
      console.log("  Adding flavor for emitter", item.key.value);
      item.value.set("flavor", "azure");
    }
  }
  if (patch) {
    const newContent = stringify(yaml, {
      aliasDuplicateObjects: false,
      lineWidth: 0,
    });
    await writeFile(file, newContent);
  }
}


```

### CODE REVIEW

File: eng/tools/typespec-validation/src/rules/flavor-azure.ts
Comment (mikeharder) [inline eng/tools/typespec-validation/src/rules/flavor-azure.ts:1]: @ckairen: Please review this new rule.  I am working on adding unit tests, will include in this PR if we have time, else in a subsequent PR.

Comment (mikeharder) [reply eng/tools/typespec-validation/src/rules/flavor-azure.ts:1]: Example failure:

```
Executing rule: FlavorAzure
"@azure-tools/typespec-python":
  flavor: undefined
"@azure-tools/typespec-java":
  flavor: undefined
"@azure-tools/typespec-csharp":
  flavor: undefined
"@azure-tools/typespec-ts":
  flavor: undefined
  
Rule FlavorAzure failed
tspconfig.yaml must define the following property:

options:
  "@azure-tools/typespec-python":
    flavor: azure

tspconfig.yaml must define the following property:

options:
  "@azure-tools/typespec-java":
    flavor: azure

tspconfig.yaml must define the following property:

options:
  "@azure-tools/typespec-csharp":
    flavor: azure

tspconfig.yaml must define the following property:

options:
  "@azure-tools/typespec-ts":
    flavor: azure
```

Comment (ckairen) [reply eng/tools/typespec-validation/src/rules/flavor-azure.ts:1]: aside from unit test, this looks good to me

**Review by ckairen — APPROVED** [review]

**Review by mikeharder — APPROVED** [review]

File: eng/tools/typespec-validation/package.json [hunk 1]
```diff
@@ -13,16 +13,17 @@
...
+        "globby": "^13.1.1",
+        "mocha": "^10.2.0",
         "typescript": "~5.3.3"
     },
     "scripts": {
-        "postinstall": "tsc",
-        "test": "mocha --recursive --exit"
+        "build": "tsc",
+        "postinstall": "npm run build",
+        "test": "npm run build && mocha --recursive --exit"
```
Comment (mikeharder) [inline eng/tools/typespec-validation/package.json:26]: Unrelated to new rule, but makes running tests easier

File: eng/tools/typespec-validation/package.json [hunk 2]
```diff
@@ -13,16 +13,17 @@
         "yaml": "^2.3.2"
     },
     "devDependencies": {
-        "globby": "^13.1.1",
-        "mocha": "^10.2.0",
         "@types/debug": "^4.1.8",
         "@types/mocha": "^10.0.2",
         "@types/node": "^18.16.18",
+        "globby": "^13.1.1",
```
Comment (mikeharder) [inline eng/tools/typespec-validation/package.json:19]: code cleanup

File: eng/tools/typespec-validation/test/flavor-azure.test.ts
Comment (mikeharder) [inline eng/tools/typespec-validation/test/flavor-azure.test.ts:1]: @ckairen: Please review added tests

File: eng/tools/typespec-validation/test/flavor-azure.test.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+  it("should succeed if config has no options", async function () {
+    let host = new TsvTestHost();
+    host.readTspConfig = async (_folder: string) => `
+emit:
+  - "@azure-tools/typespec-autorest"
+`;
+    const result = await new FlavorAzureRule().execute(host, TsvTestHost.folder);
+    assert(result.success);
+  });
```
Comment (ckairen) [inline eng/tools/typespec-validation/test/flavor-azure.test.ts:82]: Should we have a test where flavor is misplaced? (not under options: emitters:)

Comment (mikeharder) [reply eng/tools/typespec-validation/test/flavor-azure.test.ts:82]: Not sure what you mean, can you provide an example `tspconfig.yaml`?  But I don't think this is too relevant.

**Comment by weshaggard** [comment]
@lmazuel I assume all the sdk generation failures are known and expected?

**Pull request merged**